### PR TITLE
Document the MicroEJ debugger proxy

### PR DESCRIPTION
**Pull request by dalgassimou**

### CODE REVIEW

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 1]
```diff
@@ -0,0 +1,76 @@
+..  _MicroEJ-Debugger-Proxy:
+
+MicroEJ Debugger Proxy
+######################
+
+Principle
+=========
+
+The MicroEJ debugger proxy is a Java Debug Wire protocol implementation on top of MicroEJ VEE enabling on board debugging.
+It consists of a TCP server implementing the JDWP protocol and acting as a proxy between the debugger and the VEE running on the board.
```
Comment (akhabali) [inline ApplicationDeveloperGuide/debuggerProxy.rst:10]:
```suggestion
It consists of a TCP server implementing the JDWP protocol and acting as a proxy between the IDE (debugger) and the VEE (debuggee) running on the board.
```

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+.. figure:: images/debugger_proxy1.png
+   :alt: MicroEJ Development Tools Overview of the Debugger on Simulator
+   :align: center
+
+   MicroEJ Development Tools Overview of the Debugger on Simulator
```
Comment (akhabali) [inline ApplicationDeveloperGuide/debuggerProxy.rst:16]:
```suggestion
   MicroEJ Development Tools Overview of the Debugger on Board
```

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+
+.. note::
+   This feature requires Architecture version ``8.0.0`` or higher.
+
+The MicroeEJ debugger proxy only allow a postmortem debug now so it requires among other things a core dump file of a running MicroEJ VEE process.
```
Comment (akhabali) [inline ApplicationDeveloperGuide/debuggerProxy.rst:22]:
```suggestion
The current version of the VEE Debugger Proxy only supports postmortem debugging. 
It requires a core dump file of a VEE process and some libraries provided by the VEE project.
```

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 4]
```diff
@@ -0,0 +1,76 @@
+..  _MicroEJ-Debugger-Proxy:
+
+MicroEJ Debugger Proxy
```
Comment (akhabali) [inline ApplicationDeveloperGuide/debuggerProxy.rst:3]:
```suggestion
VEE Debugger Proxy
```

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+######################
+
+Principle
+=========
+
+The MicroEJ debugger proxy is a Java Debug Wire protocol implementation on top of MicroEJ VEE enabling on board debugging.
```
Comment (akhabali) [inline ApplicationDeveloperGuide/debuggerProxy.rst:9]:
```suggestion
The VEE Debugger Proxy is a Java Debug Wire Protocol (JDWP) implementation enabling on board debugging for VEE processes.
```

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+
+.. note::
+   This feature requires Architecture version ``8.0.0`` or higher.
+
+The MicroeEJ debugger proxy only allow a postmortem debug now so it requires among other things a core dump file of a running MicroEJ VEE process.
+
+Generate a coredump file for MicroEJ VEE Linux using GDB
+========================================================
+
```
Comment (akhabali) [inline ApplicationDeveloperGuide/debuggerProxy.rst:26]: In this section, I would start by explaining how the debugger can be used in production.

The coredump file should be generated whenever the VEE process receive a specific user signal or when a fatal error is raised using interrupts maybe. 

Then this core dump file can be loaded into the Debugger Proxy to perform a postmortem inspection of the VEE process state

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    echo 0x37 > /proc/self/coredump_filter
+    # Stard gdb.
+    gdb ./application.out
+    # In the gdb console run you program.
+    run
+    # When you want to generate the coredump file, suspend first the process with Ctrl-C and then generate the core dump file.
+    generate-core-file [file] # the argument `file` specifies the file name where to put the core dump
+
+
+Run the MicroEJ Debugger proxy
```
Comment (akhabali) [inline ApplicationDeveloperGuide/debuggerProxy.rst:41]:
```suggestion
Run the VEE Debugger proxy
```

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+==============================
+
+Please contact :ref:`our support team <get_support>` to get the MicroeEJ Debugger proxy program.
```
Comment (akhabali) [inline ApplicationDeveloperGuide/debuggerProxy.rst:44]:
```suggestion
Please contact :ref:`our support team <get_support>` to get the VEE Debugger Proxy program.
```

**Review by akhabali — CHANGES_REQUESTED** [review]

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+* **debugger.port**: TCP server port, default to 8000
+* **debugger.out.path**: path to VEE .out file
+* **debugger.out.coredump.path**: path to VEE .out coredump file
+* **debugger.out.format**: VEE .out file format, default to elf
+*  **debugger.out.bigEndianness**: VEE .out file big endian, default to false
+* **loader.path**: The path to the platform ``soarModelReader.jar`` and ``vmModelReader.jar`` needs to be provided here.
+  This is a Comma-separated Classpath, such as `lib,${HOME}/app/lib`.
+  Earlier entries take precedence, like a regular -classpath on the javac command line. 
+  Other runtime jars can be provided using this property.
```
Comment (amorel-microej) [inline ApplicationDeveloperGuide/debuggerProxy.rst:67]: The path must contain the name and the extension of the file (e.g. ./linux-platform/source/tools/soarModelReader.jar).

File: ApplicationDeveloperGuide/debuggerProxy.rst (hunk 5 above)
Comment (amorel-microej) [inline ApplicationDeveloperGuide/debuggerProxy.rst:9]: The MicroEJ debugger proxy is a Java Debug Wire protocol (JDWP) implementation on top of MicroEJ VEE enabling on board debugging.

File: ApplicationDeveloperGuide/debuggerProxy.rst (hunk 3 above)
Comment (amorel-microej) [inline ApplicationDeveloperGuide/debuggerProxy.rst:22]: The MicroeEJ debugger proxy only allows a postmortem debug now so it requires among other things a core dump file of a running MicroEJ VEE process.

**Review by amorel-microej — COMMENTED** [review]
It works with the generic Linux VEE on the WSL.
I tried with a Docker Linux x86 container but it doesn't work, probably the tcp access. But by get the .out and the dump file and copy its into the WSL, the process works.

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 10]
```diff
@@ -33,9 +33,35 @@ Open a shell terminal and enter the following commands:
...
-    # In the gdb console run you program.
+
+Suppose we want to generate a corefile when the the signal ``SIGUSR1`` is received or when a garbage collection (GC) is done, enter the following commands:
+
+.. code-block:: sh
+
+   # In the gdb console.
+   catch signal SIGUSR1 # catch the SIGUSR1 signal
+   break LLMJVM_on_Runtime_gc_done # break in the LLMJVM_on_Runtime_gc_done function
+   commands
```
Comment (amorel-microej) [inline ApplicationDeveloperGuide/debuggerProxy.rst:44]: The command list is for the last breakpoint or catchpoint. Here, the signal catchpoint will not generate the core dump.
It is necessary to duplicate the commands for the catchpoint.
Or to add the breakpoint or catchpoint id to the commands line.

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 11]
```diff
@@ -33,9 +33,35 @@ Open a shell terminal and enter the following commands:
...
     gdb ./application.out
-    # In the gdb console run you program.
+
+Suppose we want to generate a corefile when the the signal ``SIGUSR1`` is received or when a garbage collection (GC) is done, enter the following commands:
+
+.. code-block:: sh
+
+   # In the gdb console.
+   catch signal SIGUSR1 # catch the SIGUSR1 signal
+   break LLMJVM_on_Runtime_gc_done # break in the LLMJVM_on_Runtime_gc_done function
```
Comment (amorel-microej) [inline ApplicationDeveloperGuide/debuggerProxy.rst:43]: It can be helpful to add the symbol "LLMJVM_on_OutOfMemoryError_thrown" as an example.

**Review by akhabali — APPROVED** [review]
LGTM

**Comment by frivieremicroej** [comment]
Preview: https://docs.microej.com/en/feature-m0081meja-2056_debuggerproxy/

File: ApplicationDeveloperGuide/debug.rst [hunk 12]
```diff
@@ -3,14 +3,51 @@
 Debug an Application
 ====================
 
+Debug on Simulator
```
Comment (frivieremicroej) [inline ApplicationDeveloperGuide/debug.rst:6]: Please review & merge https://github.com/MicroEJ/docs/pull/498

File: ApplicationDeveloperGuide/debug.rst [hunk 13]
```diff
@@ -3,14 +3,51 @@
...
+
+To debug an application on Board, first run the :ref:`MicroEJ debugger proxy <MicroEJ-Debugger-Proxy>`, and run a Remote Java Application launch:
+
+- Go to :guilabel:`Run` > :guilabel:`Debug Configurations` > :guilabel:`Remote Java Application`
+  
+- Set the informations about the project to debug, the proxy connections properties, etc.
+  
+- Click on :guilabel:`Debug`
+
+.. figure:: images/debug2.png
```
Comment (frivieremicroej) [inline ApplicationDeveloperGuide/debug.rst:32]: fix crop + windows 10 screenshot (no rounded corners)

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+######################
+
+Principle
+=========
+
+The MicroEJ debugger proxy is a Java Debug Wire protocol implementation (JDWP) on top of MicroEJ VEE enabling on board debugging.
+It consists of a TCP server implementing the JDWP protocol and acting as a proxy between the IDE (debugger) and the VEE (debuggee) running on the board.
+This tool can be used to load and inspect a core dump file generated when a process crashes to get a postmortem view of the VEE state when the crache occurred.
+
+.. figure:: images/debugger_proxy1.png
```
[the remark this answers is not in the document]
Comment (frivieremicroej) [reply ApplicationDeveloperGuide/debuggerProxy.rst:13]: Then ask @frivieremicroej to update usages of VEE / terms accordingly

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+Suppose we want to generate a corefile when the the signal ``SIGUSR1`` is received, or when a out of memory error occurs, or when a garbage collection (GC) is done, enter the following commands:
+
+.. code-block:: sh
+
+   # In the gdb console.
+   catch signal SIGUSR1 # catch the SIGUSR1 signal
+   commands 
+   silent
+   generate-core-file [file] # the argument `file` specifies the file name where to put the core dump
```
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst:50]: we should replace the [file] by the filename we want

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+
+Open a shell terminal and run the following command
+
+.. code-block:: sh
+
+    java -Dloader.path=<Path to runtime jars> \
+        -Ddebugger.port=<8000> \
+        -Ddebugger.out.path=<Path to VEE .out file> \
+        -Ddebugger.out.coredump.path=<Path to VEE .out coredump file> \
+        -Ddebugger.out.format=<elf> \
```
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst:100]: https://youtrack.cross/issue/M0081MEJA-2096

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+        -Ddebugger.out.bigEndianness=<false> \
```
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst:101]: https://youtrack.cross/issue/M0081MEJA-2096

File: ApplicationDeveloperGuide/debuggerProxy.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst:14]: the schematics is an inline graph like plantuml called mermaid. do you want me to share it in text file?

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+        -jar microej-debugger-proxy.jar
+
+
+* **debugger.port**: TCP server port, default to 8000
+* **debugger.out.path**: path to VEE .out file
+* **debugger.out.coredump.path**: path to VEE .out coredump file
+* **debugger.out.format**: VEE .out file format, default to elf
+*  **debugger.out.bigEndianness**: VEE .out file big endian, default to false
```
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst:109]: https://youtrack.cross/issue/M0081MEJA-2096

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+* **loader.path**: The path to the platform ``soarModelReader.jar`` and ``vmModelReader.jar`` needs to be provided here.
```
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst:110]: https://youtrack.cross/issue/M0081MEJA-2097

File: ApplicationDeveloperGuide/debug.rst (hunk 13 above)
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debug.rst:32]: can you send the process which describe how to take a screenshot

File: ApplicationDeveloperGuide/debuggerProxy.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (frivieremicroej) [reply ApplicationDeveloperGuide/debuggerProxy.rst:14]: yes, share the source file and instructions to rebuild it

[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst:14]: where should I share the doc and the instructions?

[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst:14]: done

File: ApplicationDeveloperGuide/debuggerProxy.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst:50]: done

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+    # suspend first the process with Ctrl-C and then generate the core dump file.
+    generate-core-file [file] # the argument `file` specifies the file name where to put the core dump
+
+
+Run the VEE Debugger proxy
+==============================
+
+Please contact :ref:`our support team <get_support>` to get the VEE Debugger Proxy program.
+
+Open a shell terminal and run the following command
```
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst]: done

File: ApplicationDeveloperGuide/debug.rst (hunk 13 above)
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debug.rst:32]: done

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+.. note::
+   This feature requires Architecture version ``8.0.0`` or higher.
+
+The MicroeEJ debugger proxy only allows a postmortem debug now so it requires among other things a core dump file of a running MicroEJ VEE process.
+
+
+Generate a coredump file for MicroEJ VEE Linux using GDB
+========================================================
+
+Open a shell terminal and enter the following commands:
```
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst]: done

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+Run the program:
+
+.. code-block:: sh
+
+    # In the gdb console.
+    run
+
+
+A coredump file will be generated automatically when the ``LLMJVM_on_Runtime_gc_done`` function is called or the ``SIGUSR1`` signal is received.
+You can also suspend the process and generate the core dump file yourself:
```
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst]: done

File: ApplicationDeveloperGuide/debuggerProxy.rst [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+
+.. figure:: images/debugger_proxy1.png
+   :alt: MicroEJ Development Tools Overview of the Debugger on Board
+   :align: center
+
+   MicroEJ Development Tools Overview of the Debugger on Board
+
+
+.. note::
+   This feature requires Architecture version ``8.0.0`` or higher.
```
[the remark this answers is not in the document]
Comment (dalgassimou) [reply ApplicationDeveloperGuide/debuggerProxy.rst]: done

**Review by frivieremicroej — APPROVED** [review]

**Pull request merged**